### PR TITLE
Added new param override_file_paths

### PR DESCRIPTION
**Pull request by Aditya-2001**

### CODE REVIEW

**Review by plaird — APPROVED** [review]

File: crawl/bazel.py [hunk 1]
```diff
@@ -95,6 +96,34 @@ def query_all_libraries(repository_root_path, packages, verbose=False):
...
 
+def parse_override_file(file_path):
+    """
+    Returns a dict of {dep: overridded_dep} mapping
+    """
+    with open(file_path, "r") as f:
+        contents = f.read()
+
+        # Removes comments
+        contents = re.sub("#.*z*\n", "", contents).split("{")[1].split("}")[0]
```
Comment (kriscfoster) [inline crawl/bazel.py:107]: What is the z here for? Don't think it's needed.

**Review by kriscfoster — CHANGES_REQUESTED** [review]

File: README.md [hunk 2]
```diff
@@ -23,6 +23,19 @@ Pomgen uses the lockfiles generated by [rules_jvm_external](https://github.com/b
...
+Set `override_file_paths` parameter in [general] section of .pomgenrc file.
+
+Value of `override_file_paths` should be comma separated paths containing .bzl files and having a mapping of an dep and overridden_dep.
+
+Example `overridden.bzl`
+
+overridden = {
+  "org.sfdc-bouncycastle:bc-fips": "@org_bouncycastle//:org_bouncycastle_bc_fips",
+  "org.sfdc-bouncycastle:bctls-fips": "@org_bouncycastle//:org_bouncycastle_bctls_fips"
+}
```
Comment (kriscfoster) [inline README.md:37]: I would probably enclose this in "```" e.g.

```
overridden = {
  "org.sfdc-bouncycastle:bc-fips": "@org_bouncycastle//:org_bouncycastle_bc_fips",
  "org.sfdc-bouncycastle:bctls-fips": "@org_bouncycastle//:org_bouncycastle_bctls_fips"
}
```

File: crawl/bazel.py [hunk 3]
```diff
@@ -95,6 +96,34 @@ def query_all_libraries(repository_root_path, packages, verbose=False):
 def _normalize_dependency_string_to_group_id_artifact_id(dependency):
     return ":".join(dependency.split(":")[:2]) # remove classifier/package, so we can grab version from artifacts list
 
+def parse_override_file(file_path):
```
Comment (kriscfoster) [inline crawl/bazel.py:99]: This method could be very well unit tested e.g. testing a file with comments etc.

File: README.md [hunk 4]
```diff
@@ -23,6 +23,19 @@ Pomgen uses the lockfiles generated by [rules_jvm_external](https://github.com/b
 - pomgen < 2.0.0 is compatible with rules_jvm_external < 5.0
 - pomgen >= 2.0.0 is compatible with rules_jvm_external >= 5.0
 
+### Overidding Dependencies
```
Comment (simontoens) [inline README.md:26]: Can you see if this override file is documented in rules_jvm_external?  This section should just say something like "pomgen supports rule_jvm_external's overrides file ..." and then link to the relevant doc in rules_jvm_external.  Good to have an example here, though, keep that!

**Comment by simontoens** [comment]
I think for this feature, it would be nice to add a new directory under `examples`

File: common/overridefileinfo.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+"""
+import glob
+import os
+
+
+class OverrideFileInfo:
+
+    def __init__(self, override_file_paths):
+        self.override_file_paths = override_file_paths
```
Comment (simontoens) [inline common/overridefileinfo.py:14]: Generally, I would recommend passing the the file content into this class instead of the file path.  Then you can easily add a unit test for this class also (unit test just passes in strings, doesn't need to write files).

File: crawl/pom.py [hunk 6]
```diff
@@ -537,20 +538,41 @@ def __init__(self, workspace, artifact_def, dependency, pom_template):
...
 
+    def _update_overridden_deps(self, deps):
+        overrides_dict = self._workspace.name_to_override_dependencies
+        ext_deps = self._workspace.name_to_external_dependencies
+        output_deps = []
+        if overrides_dict == {}:
+            return deps
+        for dep in deps:
+            dep_str = dep
+            if isinstance(dep_str, dependency.ThirdPartyDependency):
```
Comment (simontoens) [inline crawl/pom.py:567]: Having to use `isinstance` in code we control usually means something is wrong.

File: crawl/workspace.py [hunk 7]
```diff
@@ -21,7 +21,7 @@ class Workspace:
     Maven concepts.
     """
     def __init__(self, repo_root_path, config, maven_install_info,
-                 pom_content, dependency_metadata, verbose=False):
+                 pom_content, dependency_metadata, override_file_info = [], verbose=False):
```
Comment (simontoens) [inline crawl/workspace.py:24]: Can't we get override_file_info from the config?  The config is passed into this ctor.

File: crawl/bazel.py [hunk 8]
```diff
@@ -95,6 +96,34 @@ def query_all_libraries(repository_root_path, packages, verbose=False):
...
+            contents = contents[:-1]
+        contents = "{" + contents.strip() + "}"
+        override_data = json.loads(contents)
+        output = {}
+
+        # Updates /, . and - with _
+        # Example - org.springframework:spring-jcl to org_springframework_spring_jcl
+        for dep, overridded_dep in override_data.items():
+            pattern = r'(?<=[^\d])\.|\.(?=[^\d])'
+            output[re.sub(pattern, "_", dep.replace(':', '_').replace("-", "_"))] = overridded_dep
```
Comment (simontoens) [inline crawl/bazel.py:124]: Should / can this logic move into the overridefileinfo class?  Then there need to be unit tests for this.

File: crawl/pom.py [hunk 9]
```diff
@@ -537,20 +538,41 @@ def __init__(self, workspace, artifact_def, dependency, pom_template):
...
         return content
 
+    def _update_overridden_deps(self, deps):
+        overrides_dict = self._workspace.name_to_override_dependencies
+        ext_deps = self._workspace.name_to_external_dependencies
+        output_deps = []
+        if overrides_dict == {}:
+            return deps
+        for dep in deps:
+            dep_str = dep
```
Comment (simontoens) [inline crawl/pom.py:566]: This is confusing.  `dep_str` would make me think it is a `str` instance, but it isn't.

File: crawl/dependency.py [hunk 10]
```diff
@@ -216,6 +216,10 @@ def bazel_label_name(self):
             name = "@%s//:%s" % (self._maven_install_name, name)
         return name
 
+    @property
+    def bazel_dep_name(self):
+        return self._bzl_artifact_name()
+
```
Comment (simontoens) [inline crawl/dependency.py:222]: Why is this needed?

File: crawl/bazel.py (hunk 1 above)
Comment (Aditya-2001) [reply crawl/bazel.py:107]: Yeah, as we discussed "#.*\n" should also work in the same way. I will try with regex once and if it works then will remove the z* part.

File: config/config.py [hunk 11]
```diff
@@ -149,6 +153,7 @@ def __str__(self):
 change_detection_enabled=%s
 """ % (self.pom_template_path_and_content[0],
        self.maven_install_paths,
+       self.override_file_paths,
```
[the remark this answers is not in the document]
Comment (Aditya-2001) [reply config/config.py:156]: I added the %s for the new param, but need to check if tests are still going through.

File: crawl/dependency.py (hunk 10 above)
Comment (Aditya-2001) [reply crawl/dependency.py:222]: To search in override file. The `bazel_label_name` property adds `_maven_install_name` to the artifact name.
Example: Override file has mapping like -> `"javax.activation:activation": "@jakarta//:jakarta_activation_jakarta_activation_api"`

`bazel_label_name` property returns `@javax//:javax.activation:activation`
'bazel_dep_name` property returns `@javax//:javax.activation:activation`

File: crawl/pom.py (hunk 6 above)
Comment (Aditya-2001) [reply crawl/pom.py:567]: There can be 2 types of dep and to fetch the correct key value to match in override file, they have diff methods.

File: common/overridefileinfo.py (hunk 5 above)
Comment (Aditya-2001) [reply common/overridefileinfo.py:14]: Yeah, that can also be done. The maveninstallinfo does the same thing i.e. takes the path. I think we can create similar testcases for this too. I will try it and write the unittest cases.

File: crawl/dependency.py (hunk 10 above)
Comment (simontoens) [reply crawl/dependency.py:222]: They don't both return the same value?

File: crawl/pom.py (hunk 6 above)
Comment (simontoens) [reply crawl/pom.py:567]: You could add a new method to the dep class and implement it in 2 diff ways in each subclass?

File: crawl/bazel.py (hunk 8 above)
Comment (Aditya-2001) [reply crawl/bazel.py:124]: Yeah, this is useful, as testing will be easy. Also, we will have minimal changes in the workspace.

File: crawl/dependency.py (hunk 10 above)
Comment (Aditya-2001) [reply crawl/dependency.py:222]: No

File: crawl/pom.py (hunk 6 above)
Comment (Aditya-2001) [reply crawl/pom.py:567]: Yeah, will do that then.

File: crawl/bazel.py (hunk 8 above)
Comment (Aditya-2001) [reply crawl/bazel.py]: Done.

File: crawl/pom.py (hunk 6 above)
Comment (Aditya-2001) [reply crawl/pom.py]: Done.

File: crawl/dependency.py (hunk 10 above)
Comment (simontoens) [reply crawl/dependency.py:226]: I mean, in the example above that you have they seem to both return the same value.

File: config/config.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Aditya-2001) [reply config/config.py:156]: The tests are running without failures.

File: README.md (hunk 4 above)
Comment (Aditya-2001) [reply README.md:26]: I have given the link to the rules_jvm_external override deps.

**Review by kriscfoster — APPROVED** [review]

File: crawl/dependency.py [hunk 12]
```diff
@@ -123,6 +123,10 @@ def references_artifact(self):
     def bazel_buildable(self):
         raise Exception("must be implemented in subclass")
 
+    @property
+    def override_matching_str(self):
+        raise Exception("must be implemented in subclass")
+
```
Comment (simontoens) [inline crawl/dependency.py:129]: Is this the value of the rhs in the "overrides file"?  Maybe this can be called override_target in that case?  Please document this attribute, see large comment on top of the file.

Comment (simontoens) [reply crawl/dependency.py:129]: Also, please add something to dependencytest for this.

File: config/config.py (hunk 11 above)
Comment (simontoens) [inline config/config.py:156]: Please add a test to configtest for this new value you added.

File: crawl/pom.py [hunk 13]
```diff
@@ -564,6 +584,9 @@ def _gen_dependencies(self, pomcontenttype):
         # account for any version overrides that need to carry over to the
         # Maven build.
         transitives = self._get_transitive_deps(self.dependencies)
+
+        # Update the dependencies according to the overridden file
+        transitives = self._update_overridden_deps(transitives)
```
Comment (simontoens) [inline crawl/pom.py:589]: Is this necessary, since the transitives are computed based on self.dependencies?

File: crawl/pom.py [hunk 14]
```diff
@@ -537,20 +538,39 @@ def __init__(self, workspace, artifact_def, dependency, pom_template):
...
 
     def gen(self, pomcontenttype):
         content = self.pom_template.replace("#{group_id}", self._artifact_def.group_id)
         content = content.replace("#{artifact_id}", self._artifact_def.artifact_id)
         version = self._artifact_def_version(pomcontenttype)
         content = content.replace("#{version}", version)
         content = self._handle_description(content, self.pom_content.description)
+
+        # Update the dependencies according to the overridden file
+        self.dependencies = self._update_overridden_deps(self.dependencies)
```
Comment (simontoens) [inline crawl/pom.py:551]: Could the update_overriden deps method be called right after parsing the pinned files, in the workspace module?

File: pomgen.py [hunk 15]
```diff
@@ -81,11 +82,13 @@ def main(args):
     if args.verbose:
         logger.debug("Global pom content: %s" % pom_content)
 
+    override_file_info = overridefileinfo.OverrideFileInfo(cfg.override_file_paths, repo_root)
     mvn_install_info = maveninstallinfo.MavenInstallInfo(cfg.maven_install_paths)
     ws = workspace.Workspace(repo_root, cfg,
```
Comment (simontoens) [inline pomgen.py:87]: Note that this is not the only place workspace is instantiated.  Also see query.py and misc/extedeps_pomgen.py

File: crawl/pom.py (hunk 13 above)
Comment (Aditya-2001) [reply crawl/pom.py:589]: Yes, the if the direct deps includes transitives which need to be overridd.en

File: crawl/pom.py (hunk 14 above)
Comment (Aditya-2001) [reply crawl/pom.py]: No, the method needs the output of the pinned deps to override the deps.

**Comment by Aditya-2001** [comment]
The check is failing due to unverified local user. Created a new pr to continue the work - https://github.com/salesforce/pomgen/pull/158.

Closing this pr.
